### PR TITLE
fix(sccm): bound opaque ids and harden finding deserialization

### PR DESCRIPTION
**Pull request by adamgell**
## What this changes

Two foundation-level hardening items in `crates/cmtraceopen-parser/src/sccm/findings.rs`, both surfaced by an exact-range CodeRabbit review of the spine as it sits on `codex/parser-family-skeleton`. Both are fail-closed tightenings; no public serialized field names change.

**1. One shared opaque-id bound (RED `10616a0e` / GREEN `ba95b4d8`).**
Before: `validate_coverage_gaps` bounded `gap.artifact_id` to 256 chars, but `validate_evidence_reference` applied no length bound at all, and `is_canonical_opaque_id` checked only non-empty and trimmed. A deserialized finding could therefore carry arbitrarily long `artifactId` and `entryId` values through `evidence`, `terminalEvidence`, and correlation-key evidence, use them as `BTreeMap` keys, and re-serialize them. `findingId`, `title`, `summary`, and the correlation-key raw and normalized fields were likewise unbounded.

After: the bound lives inside `is_canonical_opaque_id`, which structurally gates finding ids, evidence artifact and entry ids, coverage-gap artifact ids, request logical ids, and extraction profile ids. The correlation-key bound sits inside `has_canonical_value`, which every key must pass, so low-confidence keys have no bypass. Title and summary bounds follow the existing `is_bounded_request_reason` precedent exactly: trim first, count `chars()` rather than bytes.

**2. Public `Deserialize` no longer bypasses the wire contract (RED `ae622752` / GREEN `f4d30bc3`).**
Before: `SccmFindingCoverageGap` and `SccmArtifactRequest` are public and re-exported by `pub use findings::*`, and both derived `Deserialize` without `deny_unknown_fields` and without validation, while the parallel wire structs set `deny_unknown_fields` and `SccmFinding::deserialize` routes through them. Any caller deserializing those two types directly got neither strictness nor validation.

After: both keep implementing `Deserialize`, now via manual impls over the existing `deny_unknown_fields` wire structs plus `validate_coverage_gaps` / `validate_artifact_requests`. This was chosen over removing the derive because removal is a breaking API change; grep evidence that routing is safe is in the review notes below.

## Why routing rather than removing `Deserialize`

Nothing outside the finding wire path deserializes these types: no `from_value::<>`, `from_str::<>`, or `from_slice::<>` of either type anywhere in the tree; the only `Deserialize`-deriving structs naming them are the wire structs in `findings.rs`; `SccmServerIntakeAssessment` holds `Vec<SccmArtifactRequest>` but derives `Serialize` only; `src-tauri/src/` has zero `sccm` references. The change is source-compatible, and the only payloads that stop deserializing are ones the spine already considered invalid.

## Verified as already fixed, no change made

The overlapping-identity-range class (a `Users\DOMAIN\alice` value leaving `\alice` in public output) is already closed at the spine head by the merged #332 work: the sort, dedup, and merge logic is at `evidence.rs:262-271` from `59af7b5f`, and `export_merges_overlapping_identity_ranges_without_mutating_raw_snapshot` covers exactly the standalone-matcher plus user-path-matcher overlap, asserting neither `ADMIN` nor `secret` survives into exported JSON.

## Verification

- `cargo test --locked -p cmtraceopen-parser --test sccm_spine_contract`: 142 passed, 0 failed (137 baseline plus 5 new)
- `cargo test --locked -p cmtraceopen-parser`: 918 passed, 0 failed
- `cargo clippy --locked -p cmtraceopen-parser --all-targets -- -D warnings`: clean
- `cargo check --locked -p cmtraceopen-parser --target wasm32-unknown-unknown`: clean
- `git diff --check 7655826f HEAD`: clean; `rustfmt --check` clean on both changed files
- No new dependencies; `Cargo.toml` untouched; camelCase wire names preserved

## Assumptions and notes

- Bounds chosen: 256 for opaque ids, 512 for title, 2048 for summary, following the existing request-reason precedent. If the program wants different display-string limits, they are single constants.
- Nine lane branches stack on this spine, so every change here is deliberately a tightening rather than a semantic change.
- Repo note for whoever works in this crate next: a bare `cargo fmt -p cmtraceopen-parser` reformats four unrelated ESP files (`esp/redaction.rs`, `esp/reducer.rs`, `esp/timeline.rs`, `tests/esp_diagnostics.rs`) because this machine's rustfmt disagrees with their committed formatting. Format only the files you changed.

Refs #317. Not claiming native Windows acceptance; this is pure parser-layer work.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Added consistent validation for evidence references, coverage gaps, terminal evidence, artifact requests, and correlation keys.
  - Oversized, malformed, incomplete, or invalid data is now rejected instead of being accepted as valid.
  - Out-of-bounds key values are reported as malformed gaps.
  - Added safeguards for invalid line ranges, missing evidence, forged confidence values, and malformed nested data.

- **Tests**
  - Expanded coverage for boundary conditions, validation failures, malformed inputs, and successful data round trips.

### CODE REVIEW

**Comment by adamgell** [comment]
@coderabbitai full review

**Review by adamgell — COMMENTED** [review]
Exact-head review of f4d30bc3. Worktree detached at f4d30bc3, clean before and
after; all probes removed.

VERDICT: BLOCK on one item. The change is correct, non-regressive, and every
claim in the PR body is verified true. The block is scope completion on the
shared spine, not a correctness defect.

BATTERY
- cargo test --locked -p cmtraceopen-parser --test sccm_spine_contract:
  142 passed, 0 failed.
- cargo test --locked -p cmtraceopen-parser: 918 passed, 0 failed,
  14 test binaries plus doctests.
- cargo test --locked --workspace: 1672 passed, 0 failed, 30 test binaries.
- cargo clippy --locked -p cmtraceopen-parser --all-targets -- -D warnings: clean.
- cargo check --locked -p cmtraceopen-parser --target wasm32-unknown-unknown: clean.
- rustfmt --check on the two changed files only: clean.
- git diff --check 7655826f HEAD: clean.

All 12 SCCM contract binaries green, 344 tests: client deployment 8,
client health 3, client intake 3, client inventory/compliance/metering 47,
server distribution point 48, server hierarchy and replication 28,
server intake 11, server intake fixture 1, server provider and admin service 30,
server software update point 18, site core 5, spine 142.

RED DISCIPLINE, BOTH PAIRS
- RED 10616a0e: 137 passed, 3 failed. Reported mismatches included
  \"builder accepted an overlong finding ID\", \"builder accepted an overlong
  evidence artifact ID\", \"serializer accepted an overlong finding ID\", and
  \"deserializer accepted an overlong evidence artifactId\".
- GREEN ba95b4d8: 140 passed, 0 failed.
- RED ae622752: 140 passed, 2 failed. Reported mismatches included
  \"coverage gap deserializer accepted an unknown field\", \"coverage gap
  deserializer accepted captured coverage\", and \"artifact request deserializer
  accepted an undeclared logical ID\".
- GREEN f4d30bc3: 142 passed, 0 failed.

BACKWARD COMPATIBILITY
No merged lane regresses. Longest values measured across all 141 shipped
expected.json plus every fixture JSON in the corpus:

  findingId              39 / 256   (15%)
  evidence artifactId    67 / 256   (26%)
  evidence entryId       46 / 256   (18%)
  coverage gap artifactId 67 / 256  (26%)
  request logicalId      34 / 256   (13%)
  extractionProfileId    35 / 256   (14%)
  correlation key value  36 / 256   (14%)
  title                  53 / 512   (10%)
  summary                53 / 2048  (3%)
  request reason        122 / 240   (51%, bound unchanged by this PR)

Nothing is within 20% of a new limit. Two structural notes worth recording:
SccmFindingBuilder::new defaults title and summary to finding_id, so the default
path is bounded by the 256 id limit and cannot reach 512 or 2048; and
normalize_server_host self-caps at 253 characters, so the theoretical worst-case
ServerHost correlation key is 254 against a 256 limit, about 99% of the bound.
That is safe today but it is the tightest constraint in the change, and
MAX_SCCM_CORRELATION_KEY_VALUE_CHARS should not be lowered.

BOUND PROBES
Exactly-at-limit accepted and limit+1 rejected on every gated field: finding id,
evidence artifact id, evidence entry id, coverage gap artifact id, extraction
profile id, and correlation key value at 256/257; title at 512/513; summary at
2048/2049. Characters rather than bytes confirmed: 256 four-byte scalars
(1024 bytes) accepted, 257 rejected, and the value round-trips unchanged.
Trim-then-count confirmed: 512 characters wrapped in six spaces accepted and
normalized to 512, 513 wrapped in the same padding rejected, whitespace-only
rejected. Opaque ids treat padding as rejection rather than normalization, which
matches is_canonical_opaque_id. No builder versus deserialize asymmetry was found
on any gated path.

DESERIALIZE ROUTING
Both manual impls behave as claimed. Unknown fields are rejected with the wire
struct's own serde error listing the permitted names. Invalid values are rejected
by the corresponding validator. Valid payloads round-trip for both types
individually and inside a full SccmFinding carrying both. Serialized field names
are unchanged (artifactId, role, coverage; logicalId, role, reason) and
optionality is unchanged: removing a required field still produces
\"missing field\". Independent greps confirm no from_value, from_str, from_slice,
or from_reader of either type outside the test files, that
SccmServerIntakeAssessment derives Serialize only, and that src-tauri/src has no
SCCM type usage at all (its single \"SCCM\" hit is a comment string).

EVIDENCE.RS OVERLAP
Confirmed already fixed at the base, not by this PR. git blame attributes the
sort, dedup, and merge loop to 59af7b5f, which is outside this PR's four commits,
and the regression test is inside the green 918.

BLOCKING GAP

1. Three more public, re-exported SccmFinding member types still bypass the wire
   contract, and their deny_unknown_fields wire structs already exist.

   Observed at f4d30bc3 via serde_json::from_value:
   - SccmEvidenceRef accepted a 5000-character artifactId, an untrimmed entryId,
     and an unknown field.
   - SccmTerminalEvidence accepted a 5000-character nested artifactId and an
     unknown field.
   - SccmCorrelationKey accepted a 5000-character raw, an incoherent span
     (start 99999, end 1), an unknown field, and confidence \"exact\".

   The confidence case is why this is blocking rather than follow-up.
   REGISTERED_STABLE_CORRELATION_PROFILE_IDS is deliberately empty, with a comment
   stating that adding an entry requires contract review, and
   validate_correlation_key_evidence therefore requires every key to be Low.
   Standalone deserialization forges past that gate. That is a trust-signal
   bypass, materially more consequential than the length bypasses this PR closed,
   and it is exactly the class of hole the PR set out to close. Nine lanes stack
   on this spine, so closing it after they land is considerably more expensive
   than closing it now, and the remedy is the same ten-line pattern already
   applied twice in this PR: route through SccmEvidenceRefWire,
   SccmTerminalEvidenceWire, and SccmCorrelationKeyWire.

   To resolve: route those three types, or state on the PR why the spine
   deliberately stops at two.

NON-BLOCKING OBSERVATIONS

2. Collection cardinality is unbounded. A finding with 50,000 evidence refs and
   50,000 coverage gaps built, validated, serialized to roughly 8 MB, and
   deserialized cleanly. Only next_artifacts has a cardinality bound, at 16.
   The per-element strings are now bounded, but the collections are not. This is
   pre-existing and outside the stated thesis; worth a tracking issue.

3. The ServerHost correlation key headroom noted above (254 of 256) is the
   tightest ratio introduced by this change. Recommend a comment on the constant
   so it is not lowered later.

Everything else in this PR is clean. The RED/GREEN discipline is exemplary, the
bounds are placed structurally rather than at call sites, the display-string
handling follows the is_bounded_request_reason precedent exactly, and no
serialized field name or optionality changed.

**Comment by adamgell** [comment]
@coderabbitai full review

**Comment by adamgell** [comment]
### Code review
Five independent reviewers examined this change from different angles (CLAUDE.md compliance, focused bug scan, git-blame history, prior-PR feedback, and code-comment guidance). Each candidate finding was then independently verified and confidence-scored; anything below 80 is reported separately as unconfirmed.
**3 confirmed finding(s)** (score >= 80), 3 unconfirmed, 3 rejected as false positives.

#### 1. SccmCorrelationKey::deserialize leaves its nested evidence ref unvalidated — the exact door the PR set out to close
`cmtraceopen-parser/src/sccm/findings.rs` lines 404-448 (confidence 85)
Confirmed by execution, not inspection. I built a probe crate with a path dep on the branch's parser and ran the exact payloads: empty artifactId, untrimmed entryId, reversed 9->2 range, half-set range, and a 5000-char artifactId ALL deserialize successfully as a standalone SccmCorrelationKey, while the identical reference is REJECTED as a bare SccmEvidenceRef and inside SccmTerminalEvidence. Grep confirms validate_correlation_key_evidence (findings.rs:2154-2200) never calls validate_evidence_reference; its only reference check is `evidence.contains(reference)` against key.evidence.as_slice(), which is trivially self-satisfying. The sibling SccmTerminalEvidence::deserialize added in the same commit does call validate_evidence_reference at line 354, so this is an internal inconsistency in one commit, not a design choice. It directly falsifies the doc comments the PR itself adds at findings.rs:310-313 and models.rs:150-153, and bypasses the MAX_SCCM_OPAQUE_ID_CHARS bound this same PR introduces. Reachability through SccmKeyExtractionResult (models.rs:265-271, derives Deserialize) is confirmed. Test-coverage claim verified: the new correlation_key test (tests/sccm_spine_contract.rs:1863-1939) never mutates canonical["evidence"], while the terminal-evidence test at 1808-1851 does test a 5000-char nested artifactId. Held back from 100 only because nothing in-tree deserializes SccmKeyExtractionResult today outside one test, so today's exposure is latent rather than live — but on a spine nine lanes stack on, this is exactly the guarantee those lanes will assume.

Fix is one line mirroring findings.rs:354: run validate_evidence_reference over key.evidence before validate_correlation_key_evidence, or retype SccmCorrelationKeyWire.evidence as Option<SccmEvidenceRef>.

#### 2. SccmCorrelationKey::deserialize never validates its nested evidence reference — the one door the PR left open
`cmtraceopen-parser/src/sccm/findings.rs` lines 436-448 (confidence 85)
Same defect, verified. The historical framing also checks out: commit 5fd80ee9 'fix(sccm): reject nested finding wire fields' is present in this branch's history, and the wire structs it introduced carry only deny_unknown_fields, with the semantic reference checks living in SccmFinding::validate -> validate_all_evidence_references. PR #404 re-supplies that glue for the evidence-ref and terminal-evidence doors but not the correlation-key one. My probe reproduced the quoted payload verbatim (extractionProfileId "sccm-keys-experimental-v1", artifactId "", entryId untrimmed, lineStart 9 / lineEnd 2) as Ok, with both named controls rejecting the identical reference with the exact error strings quoted. SccmKeyExtractionResult reachability confirmed. Same latency caveat as the other copies keeps it below 100.

Also verified: correlation_key_deserializes_through_the_same_wire_contract_as_a_finding (tests/sccm_spine_contract.rs:1863-1939) has no nested-reference case, unlike its terminal-evidence sibling at 1808-1851.

#### 3. New standalone SccmCorrelationKey deserializer never validates its nested evidence reference
`cmtraceopen-parser/src/sccm/findings.rs` lines 406-448 (confidence 85)
Same defect, and this write-up's control matrix is the one I reproduced most exactly. My probe ran all five listed cases (5000-char artifactId, empty artifactId, untrimmed entryId, inverted 9..1 range, half-set lineStart=7/lineEnd=null) and every one was ACCEPTED as a standalone SccmCorrelationKey and through SccmKeyExtractionResult, while every one was REJECTED by both the bare SccmEvidenceRef door and the SccmTerminalEvidence door. The half-set-range case in particular is a good catch: it is the one shape a reader might assume the (None,None)/(Some,Some) span logic already covers, and it does not. All line references check out (wire field 412, impl 436-448, validator 2154-2200, terminal sibling 348-368, in-finding coverage 763-794). Below 100 only because no lane deserializes an extraction result today.

Fix: retype the wire field or add key.evidence.as_ref().map_or(Ok(()), validate_evidence_reference) before validate_correlation_key_evidence.

---

<details><summary>Unconfirmed findings (scored 50-79, verify before acting)</summary>

- **[50]** New correlation-key value bound is enforced at the validator but not at the producer, so extract_keys output no longer round-trips
- **[50]** MAX_SCCM_CORRELATION_KEY_VALUE_CHARS is enforced only at the boundary; extract_keys can now emit a key that fails the crate's own contract, taking the whole finding with it
- **[50]** New 256-char correlation-key value bound is not reconciled with the extractor that produces the values

</details>

**Comment by adamgell** [comment]
@coderabbitai full review

File: crates/cmtraceopen-parser/src/sccm/findings.rs [hunk 1]
```diff
@@ -618,9 +737,13 @@ impl SccmFindingBuilder {
...
 fn validate_required_text(finding: &SccmFinding) -> Result<(), SccmFindingValidationError> {
+    let title = finding.title.trim();
+    let summary = finding.summary.trim();
     if !is_canonical_opaque_id(&finding.finding_id)
-        || finding.title.trim().is_empty()
-        || finding.summary.trim().is_empty()
+        || title.is_empty()
+        || title.chars().count() > MAX_SCCM_FINDING_TITLE_CHARS
+        || summary.is_empty()
+        || summary.chars().count() > MAX_SCCM_FINDING_SUMMARY_CHARS
```
[the remark this answers is not in the document]
Comment (adamgell) [reply crates/cmtraceopen-parser/src/sccm/findings.rs]: Fixed in 2fc676b7. Raw stored title and summary sizes are validated before builder/deserializer normalization, so whitespace padding cannot bypass the 512/2048 limits. The regression test was observed red at builder, direct validation, serialization, and deserialization, then green; the focused spine suite is 150/150.

File: crates/cmtraceopen-parser/src/sccm/findings.rs [hunk 2]
```diff
@@ -618,9 +730,13 @@ impl SccmFindingBuilder {
...
```
[the remark this answers is not in the document]
Comment (adamgell) [reply crates/cmtraceopen-parser/src/sccm/findings.rs]: Fixed in 2fc676b7. Title, summary, and opaque-ID bounds now use one shared fail-fast character-bound helper, and raw stored title/summary are bounded before normalization. Verified by the focused 150-test spine suite, full 926-test parser suite, wasm32 check, and strict parser Clippy.

File: crates/cmtraceopen-parser/src/sccm/findings.rs [hunk 3]
```diff
@@ -2045,8 +2159,10 @@ fn validate_correlation_key_evidence(
         let normalized = normalize_key(key.kind.clone(), &key.raw);
         let has_canonical_value = !key.raw.is_empty()
             && key.raw.trim() == key.raw
+            && key.raw.chars().count() <= MAX_SCCM_CORRELATION_KEY_VALUE_CHARS
             && !key.normalized.is_empty()
             && key.normalized.trim() == key.normalized
+            && key.normalized.chars().count() <= MAX_SCCM_CORRELATION_KEY_VALUE_CHARS
```
[the remark this answers is not in the document]
Comment (adamgell) [reply crates/cmtraceopen-parser/src/sccm/findings.rs]: Fixed in 2fc676b7. Correlation-key raw and normalized bounds now use the shared fail-fast character-bound helper rather than a full character-count traversal. Existing 256/257 boundary coverage remains green; parser aggregate, wasm32, and strict Clippy also pass.

**Comment by adamgell** [comment]
### Exact-head repair pushed: 2fc676b7

Scope:
- Bounds stored title, summary, and artifact-request reason text before normalization, so whitespace padding cannot bypass the public limits.
- Replaces all full character-count bound checks in findings.rs with one fail-fast maximum-plus-one probe, including opaque IDs and correlation-key values.
- Makes standalone artifact-request serde return/emit a canonical trimmed bounded reason. Serialization intentionally does not apply finding-catalog semantics because #335 currently uses this shared transport type for server source IDs; that dependency is tracked separately.

TDD evidence:
- RED: padded display text was accepted by builder, direct validation, serialization, and deserialization.
- RED: padded request reasons were accepted by all finding paths and standalone request serde; standalone deserialization retained the noncanonical value.
- GREEN: focused SCCM spine contract is 150/150.

Verification:
- cargo test --locked -p cmtraceopen-parser --quiet: 926 passed, 0 failed.
- cargo test --locked -p cmtraceopen-parser --test sccm_server_intake: 11 passed, 0 failed.
- cargo check --locked -p cmtraceopen-parser --target wasm32-unknown-unknown: passed.
- cargo clippy --locked -p cmtraceopen-parser --all-targets -- -D warnings: passed.
- changed-file rustfmt check and git diff --check: passed.
- repository-wide cargo fmt --check --all remains red only for inherited files outside this two-file diff.

Review state:
- Local CodeRabbit 0.7.1: zero findings across both changed files.
- Replied to and resolved the three handled hosted review threads.
- Exact-head hosted Copilot and CodeRabbit review requested.

This PR remains draft. Do not merge until the new exact-head review cycle and hosted checks are green. No native Windows acceptance is claimed or needed for this pure parser-layer slice.

**Comment by adamgell** [comment]
@coderabbitai full review

Please review exact head 2fc676b7bb0e42457c93f0d64525295aa4046656. The three prior actionable threads have been fixed, replied to, and resolved; local CodeRabbit returned zero findings.

File: crates/cmtraceopen-parser/src/sccm/findings.rs [hunk 4]
```diff
@@ -168,22 +189,51 @@ impl SccmTerminalEvidence {
...
+            ));
+        }
+        SccmArtifactRequestSerializeWire {
+            logical_id: &self.logical_id,
+            role: &self.role,
+            reason,
+        }
+        .serialize(serializer)
+    }
+}
```
[the remark this answers is not in the document]
Comment (adamgell) [reply crates/cmtraceopen-parser/src/sccm/findings.rs:229]: Verified as a real contract mismatch, but the proposed full validator cannot land in this narrow slice yet. SccmServerIntakeAssessment in #335 currently serializes this shared type with server source IDs such as server-dp-distribution rather than finding-catalog logical names. Applying the proposed validator reproduced four aggregate failures (multiline parseFailed, malformed CCM, versioned provenance, and role/state matrix) with UndeclaredArtifactRequest. The dependency and two safe resolution options are tracked at https://github.com/adamgell/cmtraceopen/issues/335#issuecomment-5153692594. PR #404 preserves the existing #335 aggregate and remains draft; this thread is intentionally left unresolved until #335 stops overloading the finding request contract or maps its IDs to the catalog.

**Comment by adamgell** [comment]
### Exact-head review checkpoint

- Copilot completed a new review against 2fc676b7 and added no actionable comments.
- Hosted CodeRabbit completed the requested full review. Its one new finding is valid and is the tracked #335 request-contract dependency: https://github.com/adamgell/cmtraceopen/pull/404#discussion_r3696781131
- CodeRabbit acknowledged the dependency evidence and agreed the thread should stay unresolved until #335 stops overloading SccmArtifactRequest or maps server source IDs to catalog logical names.
- Local and remote heads match; the repair worktree is clean.

PR #404 remains draft with CHANGES_REQUESTED by design. The repair is locally green, but it is not merge-ready until #335 resolves the contract and the exact-head review cycle is rerun.

**Comment by adamgell** [comment]
Exact-head repair checkpoint for `298530cdf517b8de1217e5df5627fc603132289a`:

- Resynced non-force onto current integration `d5beb5bbb1da9c8e6b8fab22af14dee792da74c1`.
- `d7ddd228`: standalone artifact-request serialization now uses the full catalog/role/bounded-reason validator.
- `298530cd`: standalone evidence references, terminal evidence, coverage gaps, and correlation keys now validate on Serialize through the same contracts as Deserialize/containing findings.
- Added focused RED coverage for undeclared IDs, role mismatch, rooted/cross-artifact reasons, invalid/overlong evidence identities and ranges, non-terminal kinds, captured gaps, forged key confidence, missing/invalid key evidence, incoherent UTF-16 spans, and independent raw/normalized key bounds.
- Preserved camelCase wire fields and existing null/option shape.

Verification on the exact committed head:
- artifact-request focused tests: 12/12 green
- `sccm_spine_contract`: 159/159 green
- `sccm_server_intake`: 20/20 green
- `sccm_server_management_point`: 26/26 green
- full `cmtraceopen-parser`: 354 unit tests plus all integration targets and doctests green
- `cargo check --locked -p cmtraceopen-parser --target wasm32-unknown-unknown`: green
- `cargo clippy --locked -p cmtraceopen-parser --all-targets --all-features -- -D warnings`: green
- Rust 1.88 rustfmt check on all three touched files: green
- `git diff --check`: green
- repo-wide `cargo +1.88.0 fmt --all -- --check`: still red only on pre-existing ESP/Tauri paths outside this PR delta
- independent agent review: clean
- local CodeRabbit committed review of the full four-file PR delta against current integration: 0 findings

Fresh hosted CodeRabbit and Copilot reviews were requested on this exact head. The existing artifact-request thread remains unresolved until the hosted review and required CI are green.

**Comment by adamgell** [comment]
@coderabbitai full review

**Comment by adamgell** [comment]
Review follow-up pushed at exact head `a78c0288adef20595e6182f8bee02aceaa136d55`.

The CodeRabbit nitpick and Copilot suppressed observation were valid: producer-side extracted-key bounds used full `chars().count()` scans while the public validator already had a fail-fast maximum-plus-one helper. The producer now reuses that shared helper for both raw and normalized values; wire behavior and the 256-character contract are unchanged.

Verification:
- key extraction filter: 10/10 green
- SCCM spine: 159/159 green
- full parser aggregate: green
- wasm32 check: green
- strict all-target/all-feature Clippy: green
- Rust 1.88 formatting on touched files: green
- `git diff --check`: green
- local CodeRabbit on the exact two-file follow-up: 0 findings

Requesting final hosted CodeRabbit and Copilot reviews on this exact head.

**Comment by adamgell** [comment]
@coderabbitai full review
